File: src/Net/1. First Course/5. Fourth Hole/RefactoringGolf.Store/OrderItem.cs
namespace RefactoringGolf.Stack
{
    public class OrderItem
    {
        public Product Product { get; private set; }

        public int Quantity { get; private set; }

        /*
         * Order Item Constructor
         */
        public OrderItem(Product product, int quantity)
        {
            Product = product;
            Quantity = quantity;
        }

        public decimal Total()
        {
            return this.UnitPricePerQuantity() - this.CreateCategoryDiscount().CalculateDiscount(this);
        }

        private ICategoryDiscount CreateCategoryDiscount()
        {
            ICategoryDiscount categoryDiscount = null;
            if (this.Product.Category == ProductCategory.Accessories)
            {
                categoryDiscount = new AccessoriesDiscount();
            }
            if (this.Product.Category == ProductCategory.Bikes)
            {
                categoryDiscount = new BikesDiscount();
            }
            if (this.Product.Category == ProductCategory.Cloathing)
            {
                categoryDiscount = new CloathingDiscount();
            }
            return categoryDiscount;
        }

        public decimal UnitPricePerQuantity()
        {
            return this.Product.UnitPrice * this.Quantity;
        }
    }
}

File: src/Java/1. First Course/5. Fourth Hole/src/main/refactoringgolf/store/OrderItem.java
package refactoringgolf.store;

public class OrderItem {
	
	private Product product;
	private int quantity;

	/*
	 * Order Item Constructor
	 */
	public OrderItem(Product product, int quantity) {
		this.product = product;
		this.quantity = quantity;
	}
	
	public Product getProduct() {
		return product;
	}

	public int getQuantity() {
		return quantity;
	}

	public float total() {
		return itemAmount() - createCategoryDiscount().calculateDiscount(this);
	}

	private CategoryDiscount createCategoryDiscount() {
		CategoryDiscount categoryDiscount=null;
		if (getProduct().getCategory() == ProductCategory.Accessories) {
			categoryDiscount = new AccessoriesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Bikes) {
			categoryDiscount = new BikesDiscount();
		}
		if (getProduct().getCategory() == ProductCategory.Cloathing) {
			categoryDiscount = new CloathingDiscount();
		}
		return categoryDiscount;
	}

	public float itemAmount() {
		return getProduct().getUnitPrice() * getQuantity();
	}
}
